Add `samples` directory and custom session key plugins

sm-stack opened the pull request:
## Motivation

While there are basic plugins like `SingleOwnerPlugin` in `src/plugins` folder, it is not sufficient to show the different practices of plugins and the functionalities of ERC-6900. To enhance the understanding and utilization of ERC-6900 among plugin developers, establishing directories for various plugins would be beneficial.

## Solution

This involves creating two new directories: `src/samples/plugins` and `test/samples/plugins`. These directories would host a range of plugins, showcasing the different capabilities and practices within the ERC-6900 framework.

Also add a custom session key plugin inside them, which demonstrates the usage of dependency and `executeFromPluginExternal`.

jaypaik commented:
Thanks @sm-stack ! I plan on merging in the latest iteration of our address associated storage libraries soon. You'll be able to use that here instead of the PluginStorageLib currently included here. Stay tuned!

sm-stack commented:
> Update: #23

Updated some parts of the implementation with new `PluginStorageLib`!

jaypaik commented on src/samples/plugins/ModularSessionKeyPlugin.sol:
```diff
@@ -0,0 +1,338 @@
+        returns (uint256)
+    {
+        if (functionId == uint8(FunctionId.USER_OP_VALIDATION_TEMPORARY_OWNER)) {
+            (address signer,) = (userOpHash.toEthSignedMessageHash()).tryRecover(userOp.signature);
```
Nit
```suggestion
            (address signer,) = userOpHash.toEthSignedMessageHash().tryRecover(userOp.signature);
```

jaypaik commented on src/samples/plugins/ModularSessionKeyPlugin.sol:
```diff
@@ -0,0 +1,338 @@
+
+            if (_until != 0) {
+                return _packValidationData(false, _until, _after);
+            } else {
+                return _SIG_VALIDATION_FAILED;
+            }
```
Nit:
```suggestion
            }
            return _SIG_VALIDATION_FAILED;

```

jaypaik commented on src/samples/plugins/ModularSessionKeyPlugin.sol:
```diff
@@ -0,0 +1,338 @@
+                    revert WrongTimeRangeForSession();
+                }
+                return;
+            } else {
+                revert NotAuthorized();
+            }
```
```suggestion
            }
            revert NotAuthorized();
```

jaypaik commented on src/samples/plugins/ModularSessionKeyPlugin.sol:
```diff
@@ -0,0 +1,338 @@
+        manifest.executionFunctions[1] = this.removeTemporaryOwner.selector;
+        manifest.executionFunctions[2] = this.addTemporaryOwnerBatch.selector;
+        manifest.executionFunctions[3] = this.removeTemporaryOwnerBatch.selector;
+        manifest.executionFunctions[4] = this.getSessionDuration.selector;
```
It may save gas to remove this view function, since the user can call the plugin directly with the account address as an argument. Pros and cons here, but we're thinking of making a recommendation against this pattern of adding view functions to the account.

jaypaik commented on src/samples/plugins/TokenSessionKeyPlugin.sol:
```diff
@@ -0,0 +1,134 @@
+/// permitted external calls not to be changed in the future. For other child session
+/// key plugins, there can be a set of permitted external calls according to the
+/// specific needs.
+
```
```suggestion
```

jaypaik commented on src/samples/plugins/TokenSessionKeyPlugin.sol:
```diff
@@ -0,0 +1,134 @@
+
+        ManifestFunction memory tempOwnerUserOpValidationFunction = ManifestFunction({
+            functionType: ManifestAssociatedFunctionType.DEPENDENCY,
+            functionId: uint8(ISessionKeyPlugin.FunctionId.RUNTIME_VALIDATION_TEMPORARY_OWNER),
```
This should be unused since it's a dependency

jaypaik commented on test/samples/plugins/ModularSessionKeyPlugin.t.sol:
```diff
@@ -0,0 +1,381 @@
+
+contract ModularSessionKeyPluginTest is Test {
+    using ECDSA for bytes32;
+    using FunctionReferenceLib for address;
```
Nit: I feel like using `pack` this way is a bit odd. I'd prefer `FunctionReferenceLib.pack(a, b)` instead of `a.pack(b)`.

jaypaik commented on src/samples/plugins/interfaces/ISessionKeyPlugin.sol:
```diff
@@ -0,0 +1,98 @@
+    /// @param allowedSelector The selector of the function that the temporary owner is allowed to call.
+    /// @param _after The time after which the owner is valid.
+    /// @param _until The time until which the owner is valid.
+    function addTemporaryOwner(address tempOwner, bytes4 allowedSelector, uint48 _after, uint48 _until) external;
```
What do you think of calling this `addSessionKey` instead (and similar in the other functions and events)?

sm-stack replied:
That may be more intuitive, updated it.

jaypaik commented on src/samples/plugins/interfaces/ISessionKeyPlugin.sol:
```diff
@@ -0,0 +1,98 @@
+    /// @dev This function is installed on the account as part of plugin installation, and should
+    /// only be called from an account.
+    /// @param tempOwner The address of the temporary owner.
+    /// @param allowedSelector The selector of the function that the temporary owner is allowed to call.
```
It might be helpful to explain here that the intended usage pattern is to provide the function selector installed by a child session key plugin, which enforces its own permissions on the calls it can make.

jaypaik commented on src/samples/plugins/ModularSessionKeyPlugin.sol:
```diff
@@ -0,0 +1,338 @@
+import {SingleOwnerPlugin} from "../../plugins/owner/SingleOwnerPlugin.sol";
+import {PluginStorageLib, StoragePointer} from "../../libraries/PluginStorageLib.sol";
+
+/// @title Modular Session Key Plugin
+/// @author Decipher ERC-6900 Team
+/// @notice This plugin allows some designated EOA or smart contract to temporarily
+/// own a modular account.
+/// This modular session key plugin acts as a 'parent plugin' for all specific session
+/// keys. Using dependency, this plugin can be thought as a parent contract that stores
+/// session key duration information, and validation functions for session keys. All
+/// logics for session keys will be implemented in child plugins.
+/// It allows for session key owners to access MSCA both through user operation and
+/// runtime, with its own validation functions.
+/// Also, it has a dependency on SingleOwnerPlugin, to make sure that only the owner of
+/// the MSCA can add or remove session keys.
```
Could you add a note here and in the other plugin that these plugins are for demonstrating ERC-6900 capabilities only, and should not serve as examples of production-ready session key plugins?

sm-stack replied:
Added it!

sm-stack commented:
> Took a pass, thanks for your patience here! Some ERC updates in flight too, so I imagine we'll need to work through some rebases.

Addressed all feedbacks you provided, with small fixes to maintain compatibility with recent changes in the spec.

jaypaik commented on src/samples/plugins/ModularSessionKeyPlugin.sol:
```diff
@@ -0,0 +1,336 @@
+            uint48 validAfter = duration.validAfter;
+            uint48 validUntil = duration.validUntil;
+
+            if (validUntil != 0) {
+                return _packValidationData(false, validUntil, validAfter);
+            }
+            return _SIG_VALIDATION_FAILED;
```
Delaying SIG_VALIDATION_FAILED like this helps with gas estimation from bundlers when using dummy signatures.

```suggestion
            return _packValidationData(validUntil == 0, validUntil, validAfter);
```

jaypaik commented on src/samples/plugins/ModularSessionKeyPlugin.sol:
```diff
@@ -0,0 +1,336 @@
+        returns (uint256)
+    {
+        if (functionId == uint8(FunctionId.USER_OP_VALIDATION_TEMPORARY_OWNER)) {
+            (address signer,) = userOpHash.toEthSignedMessageHash().tryRecover(userOp.signature);
```
Check the result of tryRecover, and if `err != ECDSA.RecoverError.NoError`, revert with `InvalidSignature()`.

jaypaik commented on src/samples/plugins/ModularSessionKeyPlugin.sol:
```diff
@@ -0,0 +1,336 @@
+        if (functionId == uint8(FunctionId.USER_OP_VALIDATION_TEMPORARY_OWNER)) {
+            (address signer,) = userOpHash.toEthSignedMessageHash().tryRecover(userOp.signature);
+            bytes4 selector = bytes4(userOp.callData[0:4]);
+            bytes memory key = userOp.sender.allocateAssociatedStorageKey(0, 1);
```
Nit: just to be consistent with the runtime validation function

```suggestion
            bytes memory key = msg.sender.allocateAssociatedStorageKey(0, 1);
```

jaypaik commented on src/samples/plugins/ModularSessionKeyPlugin.sol:
```diff
@@ -0,0 +1,336 @@
+    {
+        if (functionId == uint8(FunctionId.RUNTIME_VALIDATION_TEMPORARY_OWNER)) {
+            bytes4 selector = bytes4(data[0:4]);
+            bytes memory key = address(msg.sender).allocateAssociatedStorageKey(0, 1);
```
Nit
```suggestion
            bytes memory key = msg.sender.allocateAssociatedStorageKey(0, 1);
```

jaypaik commented on src/samples/plugins/TokenSessionKeyPlugin.sol:
```diff
@@ -0,0 +1,133 @@
+    // ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
+
+    /// @inheritdoc ITokenSessionKeyPlugin
+    function transferFromSessionKey(address target, address from, address to, uint256 amount)
```
`target` seems redundant since only `TARGET_ERC20_CONTRACT` is valid, but if this is to showcase external call permissions, this is fine.

sm-stack replied:
Yep that was to show the wrong call to unregistered contract fails in the test code.

sm-stack commented on src/samples/plugins/ModularSessionKeyPlugin.sol:
```diff
@@ -0,0 +1,336 @@
+    }
+
+    /// @inheritdoc BasePlugin
+    function onUninstall(bytes calldata data) external override {
```
Though it will add some gas overhead, this sounds necessary to keep the consistency in the storage of plugin. I added an additional mapping that stores the information of session key and allowed selector to implement this.

sm-stack commented:
@jaypaik Addressed all the comments you suggested!

adam-alchemy reviewed (approved):
Looks good! I left a few optional suggestions, feel free to either adopt them or not.

sm-stack commented on src/samples/plugins/ModularSessionKeyPlugin.sol:
```diff
@@ -0,0 +1,379 @@
+            addr := shr(96, b32)
+            b4 := and(b32, 0xFFFFFFFF)
+        }
+    }
```
Yep that's valid! Fixed it.

sm-stack commented on src/samples/plugins/interfaces/ISessionKeyPlugin.sol:
```diff
@@ -0,0 +1,108 @@
+    /// @param _untils The times until which the owners are valid.
+    event SessionKeysAdded(
+        address indexed account,
+        address[] indexed sessionKeys,
```
I also think that this will lead to an unexpected behavior, so I just removed the indexed field at the arrays at the both events.

sm-stack commented:
@adam-alchemy Thanks for the feedbacks! Addressed to all of them.

jaypaik commented on src/samples/plugins/ModularSessionKeyPlugin.sol:
```diff
@@ -0,0 +1,379 @@
+        address account,
+        address sessionKey,
+        bytes4 allowedSelector,
+        uint48 _validAfter,
+        uint48 _validUntil
```
Nit:
```suggestion
        uint48 validAfter,
        uint48 validUntil
```

jaypaik commented on src/samples/plugins/ModularSessionKeyPlugin.sol:
```diff
@@ -0,0 +1,379 @@
+        bytes memory key = account.allocateAssociatedStorageKey(0, 1);
+        StoragePointer ptr = key.associatedStorageLookup(keccak256(abi.encodePacked(sessionKey, allowedSelector)));
+        SessionInfo storage sessionInfo = _castPtrToStruct(ptr);
+        sessionInfo.validAfter = _validAfter;
+        sessionInfo.validUntil = _validUntil;
```
```suggestion
        sessionInfo.validAfter = validAfter;
        sessionInfo.validUntil = validUntil;
```

jaypaik commented on src/samples/plugins/TokenSessionKeyPlugin.sol:
```diff
@@ -0,0 +1,133 @@
+        });
+
+        manifest.dependencyInterfaceIds = new bytes4[](2);
+        for (uint256 i = 0; i < manifest.dependencyInterfaceIds.length;) {
+            manifest.dependencyInterfaceIds[i] = type(IModularSessionKeyPlugin).interfaceId;
+            unchecked {
+                ++i;
+            }
+        }
```
Why not just:
```
manifest.dependencyInterfaceIds[0] = ...
manifest.dependencyInterfaceIds[1] = ...
```

sm-stack replied:
That's more straightforward, fixed it.

jaypaik commented on src/samples/plugins/interfaces/ISessionKeyPlugin.sol:
```diff
@@ -0,0 +1,108 @@
+    /// @param account The account whose session keys are updated.
+    /// @param sessionKeys The addresses of the session keys.
+    /// @param selectors The selectors of the functions that the session keys are allowed to call.
+    /// @param _afters The times after which the owners are valid.
+    /// @param _untils The times until which the owners are valid.
+    event SessionKeysAdded(
+        address indexed account,
+        address[] sessionKeys,
+        bytes4[] selectors,
+        uint48[] _afters,
+        uint48[] _untils
```
```suggestion
    /// @param afters The times after which the owners are valid.
    /// @param untils The times until which the owners are valid.
    event SessionKeysAdded(
        address indexed account,
        address[] sessionKeys,
        bytes4[] selectors,
        uint48[] afters,
        uint48[] untils
```

sm-stack replied:
Making `_after` to `after` leads to the following error

`Expected ',' but got reserved keyword 'after'`

So I changed all of it to `validAfter`, and same with all `until`s.

jaypaik commented on src/samples/plugins/interfaces/ISessionKeyPlugin.sol:
```diff
@@ -0,0 +1,108 @@
+    /// @param account The account whose session key is updated.
+    /// @param sessionKey The address of the session key.
+    /// @param selector The selector of the function that the session key is allowed to call.
+    /// @param _after The time after which the owner is valid.
+    /// @param _until The time until which the owner is valid.
+    event SessionKeyAdded(
+        address indexed account, address indexed sessionKey, bytes4 selector, uint48 _after, uint48 _until
+    );
```
```suggestion
    /// @param after The time after which the owner is valid.
    /// @param until The time until which the owner is valid.
    event SessionKeyAdded(
        address indexed account, address indexed sessionKey, bytes4 selector, uint48 after, uint48 until
    );
```

jaypaik commented on src/samples/plugins/interfaces/ISessionKeyPlugin.sol:
```diff
@@ -0,0 +1,108 @@
+    /// is passed as a parameter, which enforces its own permissions on the calls it can make.
+    /// @param sessionKey The address of the session key.
+    /// @param allowedSelector The selector of the function that the session key is allowed to call.
+    /// @param _after The time after which the owner is valid.
+    /// @param _until The time until which the owner is valid.
+    function addSessionKey(address sessionKey, bytes4 allowedSelector, uint48 _after, uint48 _until) external;
```
```suggestion
    /// @param after The time after which the owner is valid.
    /// @param until The time until which the owner is valid.
    function addSessionKey(address sessionKey, bytes4 allowedSelector, uint48 after, uint48 until) external;
```

jaypaik commented on src/samples/plugins/interfaces/ISessionKeyPlugin.sol:
```diff
@@ -0,0 +1,108 @@
+    /// only be called from an account.
+    /// @param sessionKeys The addresses of the session keys.
+    /// @param allowedSelectors The selectors of the functions that the session keys are allowed to call.
+    /// @param _afters The times after which the owners are valid.
+    /// @param _untils The times until which the owners are valid.
+    function addSessionKeyBatch(
+        address[] calldata sessionKeys,
+        bytes4[] calldata allowedSelectors,
+        uint48[] calldata _afters,
+        uint48[] calldata _untils
```
```suggestion
    /// @param afters The times after which the owners are valid.
    /// @param untils The times until which the owners are valid.
    function addSessionKeyBatch(
        address[] calldata sessionKeys,
        bytes4[] calldata allowedSelectors,
        uint48[] calldata afters,
        uint48[] calldata untils
```

jaypaik reviewed (commented):
Some final nits!

Also noticed you use `unchecked { ++i; }` in some places but not others. Would prefer those be consistent too!

jaypaik reviewed (approved):
🥳

jaypaik commented:
Looks like lint failed, can you run forge fmt?

sm-stack commented:
Forgot it, just committed :) @jaypaik
